In [1]:
# Import Libraries.
import nltk
from nltk import word_tokenize
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from nltk.corpus import sentiwordnet as swn
from nltk.sentiment.vader import SentimentIntensityAnalyzer

import time

/home/lorrel_plimier/anaconda3/lib/python3.7/site-packages/nltk/twitter/__init__.py:20: UserWarning: The twython library has not been installed. Some functionality from the twitter package will not be available.
  warnings.warn("The twython library has not been installed. "


In [2]:
annotations = pd.read_csv("../data/fb_annotations.csv")
annotations

source op_gender  \
0     facebook_congress         M   
1     facebook_congress         M   
2     facebook_congress         M   
3     facebook_congress         M   
4     facebook_congress         M   
5     facebook_congress         M   
6     facebook_congress         M   
7     facebook_congress         M   
8     facebook_congress         M   
9     facebook_congress         M   
10    facebook_congress         M   
11    facebook_congress         M   
12    facebook_congress         M   
13    facebook_congress         M   
14    facebook_congress         M   
15    facebook_congress         M   
16    facebook_congress         M   
17    facebook_congress         M   
18    facebook_congress         M   
19    facebook_congress         M   
20    facebook_congress         M   
21    facebook_congress         M   
22    facebook_congress         M   
23    facebook_congress         M   
24    facebook_congress         M   
25    facebook_congress         M   
26    facebook_congress         M   
27    facebook_congress         M   
28    facebook_congress         M   
29    facebook_congress         M   
...                 ...       ...   
4086  facebook_congress         M   
4087  facebook_congress         M   
4088  facebook_congress         M   
4089  facebook_congress         M   
4090  facebook_congress         M   
4091  facebook_congress         M   
4092  facebook_congress         M   
4093  facebook_congress         M   
4094  facebook_congress         M   
4095  facebook_congress         M   
4096  facebook_congress         M   
4097  facebook_congress         M   
4098  facebook_congress         M   
4099  facebook_congress         M   
4100  facebook_congress         M   
4101  facebook_congress         M   
4102  facebook_congress         M   
4103  facebook_congress         M   
4104  facebook_congress         M   
4105  facebook_congress         M   
4106  facebook_congress         M   
4107  facebook_congress         M   
4108  facebook_congress         M   
4109  facebook_congress         M   
4110  facebook_congress         M   
4111  facebook_congress         M   
4112  facebook_congress         M   
4113  facebook_congress         M   
4114  facebook_congress         M   
4115  facebook_congress         M   

                                              post_text  \
0     We're keeping the victims and families of toda...   
1     We're keeping the victims and families of toda...   
2     I hope everybody has a great weekend! Let's ta...   
3     I hope everybody has a great weekend! Let's ta...   
4     In Albany, Texas tonight at the 15th Annual We...   
5     In Albany, Texas tonight at the 15th Annual We...   
6     Had a great time speaking to many passionate p...   
7     Had a great time speaking to many passionate p...   
8     Tonight I enjoyed the opening ceremony of the ...   
9     Thanks to Jim Kelley and the employees at Sola...   
10    Exciting things are happening on Market Street...   
11    The numbers from the 2015 health care insuranc...   
12    Happy 239th Birthday to the Marine Corps! And ...   
13    I'm so happy that the I495 Bridge over the Chr...   
14    It was great being a part of Lewes' Historical...   
15    I am proud to announce that the President's ne...   
16    Today, we celebrate the 150th anniversary of P...   
17    Open Enrollment for Medicare is happening now ...   
18    Today I stopped by United Electric before jump...   
19                 These students have a bright future!   
20    I'm a strong 2nd Amendment supporter -- my fam...   
21    On Monday I joined Gov. Jack Markell and Delaw...   
22    I'm heading down to the Sussex County Emergenc...   
23    Today, I enjoyed visiting with a couple dozen ...   
24    Tonight, President Obama laid out a positive v...   
25    Sen. Carper receives a perfect 100% score on L...   
26    Go Blue Hens! Good luck in tonight's national ...   
27    Delaware's Congressional Delegation Announces ...   
28    Sen. 

In [6]:
annot_nans = annotations[annotations.isna().any(axis=1)]
annot_nans

Empty DataFrame
Columns: [source, op_gender, post_text, response_text, sentiment, relevance, op_id, post_id, post_type]
Index: []

In [3]:
# Run Vader sentiment analysis on each response in the annotations set.
start_time = time.time()
sia = SentimentIntensityAnalyzer()
Vader_sent = annotations["response_text"].apply(sia.polarity_scores)
Vader_sent = [d.get('compound') for d in Vader_sent]
annotations['Vader_sent'] = Vader_sent
print("run time:",time.time()-start_time)
annotations

run time: 1.0908098220825195


source op_gender  \
0     facebook_congress         M   
1     facebook_congress         M   
2     facebook_congress         M   
3     facebook_congress         M   
4     facebook_congress         M   
5     facebook_congress         M   
6     facebook_congress         M   
7     facebook_congress         M   
8     facebook_congress         M   
9     facebook_congress         M   
10    facebook_congress         M   
11    facebook_congress         M   
12    facebook_congress         M   
13    facebook_congress         M   
14    facebook_congress         M   
15    facebook_congress         M   
16    facebook_congress         M   
17    facebook_congress         M   
18    facebook_congress         M   
19    facebook_congress         M   
20    facebook_congress         M   
21    facebook_congress         M   
22    facebook_congress         M   
23    facebook_congress         M   
24    facebook_congress         M   
25    facebook_congress         M   
26    facebook_congress         M   
27    facebook_congress         M   
28    facebook_congress         M   
29    facebook_congress         M   
...                 ...       ...   
4086  facebook_congress         M   
4087  facebook_congress         M   
4088  facebook_congress         M   
4089  facebook_congress         M   
4090  facebook_congress         M   
4091  facebook_congress         M   
4092  facebook_congress         M   
4093  facebook_congress         M   
4094  facebook_congress         M   
4095  facebook_congress         M   
4096  facebook_congress         M   
4097  facebook_congress         M   
4098  facebook_congress         M   
4099  facebook_congress         M   
4100  facebook_congress         M   
4101  facebook_congress         M   
4102  facebook_congress         M   
4103  facebook_congress         M   
4104  facebook_congress         M   
4105  facebook_congress         M   
4106  facebook_congress         M   
4107  facebook_congress         M   
4108  facebook_congress         M   
4109  facebook_congress         M   
4110  facebook_congress         M   
4111  facebook_congress         M   
4112  facebook_congress         M   
4113  facebook_congress         M   
4114  facebook_congress         M   
4115  facebook_congress         M   

                                              post_text  \
0     We're keeping the victims and families of toda...   
1     We're keeping the victims and families of toda...   
2     I hope everybody has a great weekend! Let's ta...   
3     I hope everybody has a great weekend! Let's ta...   
4     In Albany, Texas tonight at the 15th Annual We...   
5     In Albany, Texas tonight at the 15th Annual We...   
6     Had a great time speaking to many passionate p...   
7     Had a great time speaking to many passionate p...   
8     Tonight I enjoyed the opening ceremony of the ...   
9     Thanks to Jim Kelley and the employees at Sola...   
10    Exciting things are happening on Market Street...   
11    The numbers from the 2015 health care insuranc...   
12    Happy 239th Birthday to the Marine Corps! And ...   
13    I'm so happy that the I495 Bridge over the Chr...   
14    It was great being a part of Lewes' Historical...   
15    I am proud to announce that the President's ne...   
16    Today, we celebrate the 150th anniversary of P...   
17    Open Enrollment for Medicare is happening now ...   
18    Today I stopped by United Electric before jump...   
19                 These students have a bright future!   
20    I'm a strong 2nd Amendment supporter -- my fam...   
21    On Monday I joined Gov. Jack Markell and Delaw...   
22    I'm heading down to the Sussex County Emergenc...   
23    Today, I enjoyed visiting with a couple dozen ...   
24    Tonight, President Obama laid out a positive v...   
25    Sen. Carper receives a perfect 100% score on L...   
26    Go Blue Hens! Good luck in tonight's national ...   
27    Delaware's Congressional Delegation Announces ...   
28    Sen. 

In [4]:
def convert_sentiment(x):
    if x == 'Negative':
        return -1
    elif x == 'Positive':
        return 1
    else:
        return 0

In [5]:
def convert_Vader(x):
    if x < 0:
        return -1
    elif x == 0:
        return 0
    else:
        return 1

In [6]:
annotations['Vader_sent_01'] = annotations['Vader_sent'].apply(convert_Vader)
annotations['sent_01'] = annotations['sentiment'].apply(convert_sentiment)
annotations

source op_gender  \
0     facebook_congress         M   
1     facebook_congress         M   
2     facebook_congress         M   
3     facebook_congress         M   
4     facebook_congress         M   
5     facebook_congress         M   
6     facebook_congress         M   
7     facebook_congress         M   
8     facebook_congress         M   
9     facebook_congress         M   
10    facebook_congress         M   
11    facebook_congress         M   
12    facebook_congress         M   
13    facebook_congress         M   
14    facebook_congress         M   
15    facebook_congress         M   
16    facebook_congress         M   
17    facebook_congress         M   
18    facebook_congress         M   
19    facebook_congress         M   
20    facebook_congress         M   
21    facebook_congress         M   
22    facebook_congress         M   
23    facebook_congress         M   
24    facebook_congress         M   
25    facebook_congress         M   
26    facebook_congress         M   
27    facebook_congress         M   
28    facebook_congress         M   
29    facebook_congress         M   
...                 ...       ...   
4086  facebook_congress         M   
4087  facebook_congress         M   
4088  facebook_congress         M   
4089  facebook_congress         M   
4090  facebook_congress         M   
4091  facebook_congress         M   
4092  facebook_congress         M   
4093  facebook_congress         M   
4094  facebook_congress         M   
4095  facebook_congress         M   
4096  facebook_congress         M   
4097  facebook_congress         M   
4098  facebook_congress         M   
4099  facebook_congress         M   
4100  facebook_congress         M   
4101  facebook_congress         M   
4102  facebook_congress         M   
4103  facebook_congress         M   
4104  facebook_congress         M   
4105  facebook_congress         M   
4106  facebook_congress         M   
4107  facebook_congress         M   
4108  facebook_congress         M   
4109  facebook_congress         M   
4110  facebook_congress         M   
4111  facebook_congress         M   
4112  facebook_congress         M   
4113  facebook_congress         M   
4114  facebook_congress         M   
4115  facebook_congress         M   

                                              post_text  \
0     We're keeping the victims and families of toda...   
1     We're keeping the victims and families of toda...   
2     I hope everybody has a great weekend! Let's ta...   
3     I hope everybody has a great weekend! Let's ta...   
4     In Albany, Texas tonight at the 15th Annual We...   
5     In Albany, Texas tonight at the 15th Annual We...   
6     Had a great time speaking to many passionate p...   
7     Had a great time speaking to many passionate p...   
8     Tonight I enjoyed the opening ceremony of the ...   
9     Thanks to Jim Kelley and the employees at Sola...   
10    Exciting things are happening on Market Street...   
11    The numbers from the 2015 health care insuranc...   
12    Happy 239th Birthday to the Marine Corps! And ...   
13    I'm so happy that the I495 Bridge over the Chr...   
14    It was great being a part of Lewes' Historical...   
15    I am proud to announce that the President's ne...   
16    Today, we celebrate the 150th anniversary of P...   
17    Open Enrollment for Medicare is happening now ...   
18    Today I stopped by United Electric before jump...   
19                 These students have a bright future!   
20    I'm a strong 2nd Amendment supporter -- my fam...   
21    On Monday I joined Gov. Jack Markell and Delaw...   
22    I'm heading down to the Sussex County Emergenc...   
23    Today, I enjoyed visiting with a couple dozen ...   
24    Tonight, President Obama laid out a positive v...   
25    Sen. Carper receives a perfect 100% score on L...   
26    Go Blue Hens! Good luck in tonight's national ...   
27    Delaware's Congressional Delegation Announces ...   
28    Sen. 

In [7]:
V_comp = annotations['sent_01'] == annotations['Vader_sent_01']
V_comp.value_counts()

True     2363
False    1753
dtype: int64

In [13]:
2363/len(annotations)

0.5741010689990281

In [11]:
# Get a random sample of the varios mismatched sentiments
# Negative annotation, neutral Vader:
annot_neg_neu = annotations[(annotations['sent_01']== -1) & (annotations['Vader_sent_01']==0)]
annot_neg_neu.reset_index(drop=True,inplace=True)
i = np.random.randint(len(annot_neg_neu),size=5)
comparison1 = annot_neg_neu.iloc[i][['response_text','sent_01','Vader_sent_01']]
# Negative annotation, positive Vader:
annot_neg_pos = annotations[(annotations['sent_01']== -1) & (annotations['Vader_sent_01']==1)]
annot_neg_pos.reset_index(drop=True,inplace=True)
i = np.random.randint(len(annot_neg_pos),size=5)
comparison2 = annot_neg_pos.iloc[i][['response_text','sent_01','Vader_sent_01']]
# neutral annotation, negative Vader:
annot_neu_neg = annotations[(annotations['sent_01']== 0) & (annotations['Vader_sent_01']==-1)]
annot_neu_neg.reset_index(drop=True,inplace=True)
i = np.random.randint(len(annot_neu_neg),size=5)
comparison3 = annot_neu_neg.iloc[i][['response_text','sent_01','Vader_sent_01']]
# Neutral annotation, positive Vader:
annot_neu_pos = annotations[(annotations['sent_01']== 0) & (annotations['Vader_sent_01']==1)]
annot_neu_pos.reset_index(drop=True,inplace=True)
i = np.random.randint(len(annot_neu_pos),size=5)
comparison4 = annot_neu_pos.iloc[i][['response_text','sent_01','Vader_sent_01']]
# Positive annotation, negative Vader:
annot_pos_neg = annotations[(annotations['sent_01']== 1) & (annotations['Vader_sent_01']==-1)]
annot_pos_neg.reset_index(drop=True,inplace=True)
i = np.random.randint(len(annot_pos_neg),size=5)
comparison5 = annot_pos_neg.iloc[i][['response_text','sent_01','Vader_sent_01']]
# Positive annotation, neutral Vader:
annot_pos_neu = annotations[(annotations['sent_01']== 1) & (annotations['Vader_sent_01']==0)]
annot_pos_neu.reset_index(drop=True,inplace=True)
i = np.random.randint(len(annot_pos_neu),size=5)
comparison6 = annot_pos_neu.iloc[i][['response_text','sent_01','Vader_sent_01']]

In [19]:
# Combine the above into one dataframe.
comparisons = pd.concat([comparison,comparison2,comparison3,comparison4,
                         comparison5,comparison6])

response_text  sent_01  Vader_sent_01
126  So now next is money as a right, an equal righ...       -1              0
35                 @Kirk...and the have and have nots!       -1              0
165                 I'd rather have a root canal . . .       -1              0
174   Won't any legitimate news program interview you?       -1              0
187  I did not get paid to show up at an open house...       -1              0
77   And will you support the Medical Marijuana law...       -1              1
471  I am happy to say that Tim's first official po...       -1              1
467  If you truly believe this ..stop voting agains...       -1              1
114  and for god's sakes people, if there's a butte...       -1              1
328  I guess it doesn't surprise me that a leftist ...       -1              1
60   sorry I missed you in Fredericksburg on Saturd...        0             -1
149  Question? so who cut the Soldiers benifit? Gua...        0             -1
63   with your children....haha...couldn't help mys...        0             -1
44   SAYING THINGS IN ALL CAPITALS DOES NOT MAKE TH...        0             -1
58   Ignorant sheep believing these lies about the ...        0             -1
19             A look like business & gentlewoman&man.        0              1
256  find some educators if you want to hear about ...        0              1
39   All this attention from the DCCC makes me want...        0              1
441  Abe Lincoln challenges you to do the Ice Bucke...        0              1
497  I'll bet he was fiery. Always fun to see peopl...        0              1
71   Also quite another to lose Friday night and th...        1             -1
10   that is just a horrible thing to happen. Sad s...        1             -1
71   Also quite another to lose Friday night and th...        1             -1
45   Never did "like" Mater Dei back in the good ol...        1             -1
7    Did they explain how much they dump in the Riv...        1             -1
59                     did that already. you got mine.        1              0
34           We need to keep Bob Menendez in congress.        1              0
14   We are Voting early October 22 in Las Vegas fo...        1              0
8        Getmade Monique voted...and I will vote again        1              0
71   First two at our polling place they opened at ...        1              0

In [21]:
# Print the full text of the responses with the conflicting sentiment values.
comparisons.reset_index(drop=True,inplace=True)
for i in range(len(comparisons)):
    print(comparisons.iloc[i]['response_text'])
    print('annotation sentiment:',comparisons.iloc[i]['sent_01'])
    print('Vader sentiment',comparisons.iloc[i]['Vader_sent_01'])
    print('-----------------')
    

So now next is money as a right, an equal right. Take it from those who have it, give to those who don't, and we're all equal, right?
annotation sentiment: -1
Vader sentiment 0
-----------------
@Kirk...and the have and have nots!
annotation sentiment: -1
Vader sentiment 0
-----------------
I'd rather have a root canal . . .
annotation sentiment: -1
Vader sentiment 0
-----------------
Won't any legitimate news program interview you?
annotation sentiment: -1
Vader sentiment 0
-----------------
I did not get paid to show up at an open house that was not an open house. And I didn't get paid to stand outside in the cold!
annotation sentiment: -1
Vader sentiment 0
-----------------
And will you support the Medical Marijuana laws of Hawai'i, and the thousands of patients who find relief not attainable with pharmceutical prescriptions?
annotation sentiment: -1
Vader sentiment 1
-----------------
I am happy to say that Tim's first official political act as an 18 year old will be to sign your n

### Conclusions

Based on this random sampling, the Vader sentiment is often more accurate than the human entered annotation though at times the opposite is also true. Occasionally the resonse is ambiguous and could be either sentiment. Although our overall accuracy for this sentiment analysis is 58%, this is only in comparison to what appears to be an unrealiable rating from the Mechanical Turk human readers.

Below are a few examples of discrepancies:

|Response   	|  Vader 	|  Human 	|
|---	|---	|---	|
|  that is just a horrible thing to happen. Sad statement of the times we are living. 	|   -1	|   1	|
|  All this attention from the DCCC makes me want to blush!|   1	|   0	|
|Ignorant sheep believing these lies about the Ryn budget without even reading the proposals make me frigging sick! Ask Debbie where the Senate dems budget has beeen for going on 4 years?|   -1	|   0	|